### DISCUSSION TITLE
v0.1.0-alpha: Feedback

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/eslint-config#55] [Releases]
@Krakazybik I suggest to collect here known issues after using 0.1.0 prototype versions (for quick fix)

**Comment by azinit** [comment]
**Problem:**  Not working linting of private segments
**Rules:** `public-api-boundaries`
**Version:** `0.0.0-concept.2`
**Reference:** https://github.com/ani-team/github-client/blob/workshop/feature-sliced-next/src/widgets/header/index.tsx
**Issue:** #53 

![image](https://user-images.githubusercontent.com/42924400/147039336-1331bd71-d839-44e3-b699-3cf7393e67b0.png)

**Reply by Krakazybik** [reply]
проблема связана с доступом по пути features/repo-search/ui.tsx

**Reply by Krakazybik** [reply]
Fixed in #54

**Comment by azinit** [comment]
**Problem:** False trigger on valid imports 
**Rules:** `public-api`
**Version:** `0.1.0-alpha`
**Issue:** https://github.com/feature-sliced/eslint-config/issues/60

![image](https://user-images.githubusercontent.com/42924400/147461248-ccb38125-de96-4727-86c1-c4f8a302b889.png)

**Reply by Krakazybik** [reply]
Try to fix in #61

**Comment by azinit** [comment]
**Problem:** Incorrect processing of aliased-imports
**Rules:** `import-order`
**Version:** `0.1.0-alpha.1`
**Reference:** https://github.com/tednaaa/paint
**Issue:** #58 

<details>
<summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/42924400/147757166-ea74f109-0aa6-447d-b9bd-8b906b17efc1.png)

![image](https://user-images.githubusercontent.com/42924400/147757234-45c1945a-4c76-49ca-aa90-eaf8b4e48436.png)

![image](https://user-images.githubusercontent.com/42924400/147757267-266a1219-2659-49f8-936b-279b234f3b15.png)

![image](https://user-images.githubusercontent.com/42924400/147757286-142bd787-10ea-42a9-aa92-73ef0c50ff95.png)

![image](https://user-images.githubusercontent.com/42924400/147757306-7c4154dd-1f59-4a24-9c5d-9a49ef426e59.png)

</details>

**Reply by azinit** [reply]
<details>

<summary>Config for repro</summary>

```json
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
      "ecmaVersion": 2020,
      "ecmaFeatures": {
          "jsx": true,
          "modules": true
      },
      "sourceType": "module"
  },
  "extends": [
    // "alloy", 
    // "alloy/react", 
    // "alloy/typescript",
    "@feature-sliced/eslint-config/rules/import-order"
    // "@feature-sliced/eslint-config/rules/public-api",
    // "@feature-sliced/eslint-config/rules/layers-slices",
  ],
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "typescript": {
          "alwaysTryTypes": true
      }
    }
  },
  "env": {
    "jest": true,
    "browser": true,
    "es6": true
  },
  "rules": {
    "no-require-imports": 0
  }
}
```
</details>

**Reply by azinit** [reply]
Other examples: https://github.com/illright/draw-ill-help

![image](https://user-images.githubusercontent.com/42924400/147763110-e030ac5a-323a-470e-b021-33de15c9b303.png)

![image](https://user-images.githubusercontent.com/42924400/147763143-b2f98d95-9f1e-498b-a235-55e9ca21c567.png)

**Reply by Krakazybik** [reply]
partial fix in #64

**Reply by azinit** [reply]
А поч "частично"?)
Разве поддержка альясов не налаживает корректный порядок? Или не тестил еще?

**Reply by Krakazybik** [reply]
Ну частично, из-за того, что мы решили сделать поддержку только для корневых альясов, аля "@" "@root" "~". А альясы вроде "@lib => /src/shared/lib" решили не реализовывать, т.к. это потребует совсем другую реализацию.

**Reply by azinit** [reply]
А, ну это то да)

**Comment by azinit** [comment]
**Problem:** Falsy triggered on imports of local modules
**Rules:** `public-api`
**Version:** `0.1.0-alpha.1`
**Reference:** https://github.com/tednaaa/paint

![image](https://user-images.githubusercontent.com/42924400/147757607-1d49301c-35a3-4790-abfc-a6efb0ca10c7.png)

**Reply by azinit** [reply]
![image](https://user-images.githubusercontent.com/42924400/147757892-87b4e04d-5806-4692-a1de-4f6d14b46d61.png)

**Reply by azinit** [reply]
В целом то он верно с какой-то стороны отлавливает

Но кажется реэкспортить внутренности на уровне сегментов - вполне легально
(в идеале такой потом отдельным правилом линтить)

![image](https://user-images.githubusercontent.com/42924400/147758464-4838521e-620d-49a3-9c20-f54c24384fbb.png)

**Reply by Krakazybik** [reply]
Хм.. Как мне **кажется** не стоит делать такие послабления на уровне basic eslint-config'a, т.к. **eslint** - это всё таки _набор бэст практикс_.

_Это по сути отключение обязательности public-api (т.е. отключение правила), даже из скринов мы можем видет что изменнение названия tool.model - скорее всего потянет за собой изменеение во всех интруменах. Нарушает ли это требования (а также практик SOLID/etc) устойчивости к изменениям PublicAPI? Мне кажется, отчасти, да. Является ли это бэст практикс? Я хочу от тебя услышать =)_


Методолгия предоставляет множество возможностей и старается по минимумы ограничивать и многие вещи выглядят как _"так лучше не делать, но делать можно"  (и я не уверен, что все  эйдж кейсы мы сможем завернуть в конфиг)_, **_но, как мне кажется, это не про _eslint-config_ , тем более ты сам знаешь о его возможностях =)_**

**Врядли мы сможем подстроиться под всех на данном этапе и с текущими возможностями конфига.**
И да, возможно в дальнейшем, на уровне плагина, стоит ввести возможность кастомизации доступа для Public.

**Reply by azinit** [reply]
> т.к. **eslint** - это всё таки _набор бэст практикс_.

> _"так лучше не делать, но делать можно" (и я не уверен, что все эйдж кейсы мы сможем завернуть в конфиг)_, **_но, как мне кажется, это не про _eslint-config_ , тем более ты сам знаешь о его возможностях =)_**

Да, тут история скорее про то, что public-api, как и любая концепция FSD, имеет несколько степеней применимости согласно строгости/профитности

1) `public-api#slices` - реэкспорты на уровне слайсов обязательны
2) `public-api#segments` - + реэкспорты на уровне сегментов обязательны
3) `public-api#absolute` - + реэкспорты обязательны везде

Тут речь как раз про то, что хоть eslint-config и может линтить сразу на уровне 3, но люди еще адаптируются и пытаются принять методологию только на уровне 1

А если и eslint их настолько блочит, то могут быть проблемы (хотя про бест практис с твоей стороны - тоже резонно в каком-то виде)

**Reply by azinit** [reply]
> **Врядли мы сможем подстроиться под всех на данном этапе и с текущими возможностями конфига.**
> И да, возможно в дальнейшем, на уровне плагина, стоит ввести возможность кастомизации доступа для Public.

Согласен, и тут скорее "замечание на будущее" нам

Т.к. сплитить щас эти уровни кажется рили будет проблемным, но важным позднее

**Comment by azinit** [comment]
**Problem:** Shared should be able to import ANY other shared-segments
**Rules:** `layers-slices`
**Version:** `0.1.0-alpha.1`
**Reference:** https://github.com/tednaaa/paint

![image](https://user-images.githubusercontent.com/42924400/147758715-5f7525d3-1e5a-414b-90f5-a8619c963d92.png)

**Reply by azinit** [reply]
Позже наверное еще добавим правило `segments-boundaries`, чтобы соблюдался порядок:

> `[ANY_CUSTOM_SEGMENT]` => `ui` => `model` => `api` => `lib` => `config`

Для shared тоже будет работать, но ща надо разрешить shared импортить любой ресурс из shared

**Reply by Krakazybik** [reply]
так и не добавили shared сегменты

**Reply by azinit** [reply]
> так и не добавили shared сегменты

@Krakazybik Ну вот надо бы добавить этот слой как исключение)

**Reply by Krakazybik** [reply]
Fixed in #65

**Comment by azinit** [comment]
**Problem:** Is no supported for .svelte files
**Rules:** `import-order`
**Version:** `0.1.0-alpha.1`
**Reference:** https://github.com/illright/draw-ill-help

![image](https://user-images.githubusercontent.com/42924400/147762991-29d9f707-3ca0-4ba1-b855-e3b58a371ef9.png)


![image](https://user-images.githubusercontent.com/42924400/147762957-ec9dd332-e6d3-43f8-8ba4-62d7720b77ff.png)

**Reply by azinit** [reply]
И как полагаю - для `.vue` скорее всего тоже. Но точнее - позже скажу

**Reply by azinit** [reply]
Хотя для каких то работает 🤔 

![image](https://user-images.githubusercontent.com/42924400/147763196-fd5f7bc1-b6ab-4244-a2d3-e55cfc4aa079.png)
